feat(state): introduce actions as a function (#1585)
rx-angular/rx-angular: apps/docs/docs/state/actions/actions.mdx

[![npm](https://img.shields.io/npm/v/%40rx-angular%2Fstate.svg)](https://www.npmjs.com/package/%40rx-angular%2Fstate)
![rx-angular CI](https://github.com/rx-angular/rx-angular/workflows/rx-angular%20CI/badge.svg?branch=master)

> > This package provides a handy way handle have reactive events of any type.
> > It fit's perfectly for situations like debounce user input or switchMap a HTTP request

Reviewer: ```suggestion
> `RxActions` is a powerful yet simple tool to manage actions of any type.
> It can be seen as the glue between user based events and your applications state. 
```

## Key features

- ✅ No-Boilerplate
- ✅ Minimal memory footprint through a Proxy object and lazy initialization

> - ✅ Auto unsubscribe on component or service destroy
> - ✅ Trigger effects
> - ✅ Configurable transformations to have lines in the template
> - ✅ Fully Typed

Reviewer: ```suggestion
- ✅ Automatic subscription handling
- ✅ Dispatch actions
- ✅ Transformations help you to have a clean template
- ✅ Fully Typed
```

yarn add @rx-angular/state
```

This package helps to reduce code used to create composable action streams.
It mostly is used in combination with state management libs to handle user interaction and backend communication.

> ## Setup

Reviewer: ```suggestion
## Usage
```

> `Actions` can be imported as following:

Reviewer: ```suggestion
`RxActions` are instantiated by the creation function `rxActions` coming from the `@rx-angular/state/actions` package.
```

```ts
import { rxActions } from '@rx-angular/state/actions';

@Component({
  standalone: true,
  template: `...`,
})
export class AnyComponent {
  actions = rxActions<UI>();
}
```

### Basic usage in a Component

> # Actions

Reviewer: ```suggestion

```

why do we have this headline? Does it make sense? Looks like we can just remove it

> # Actions

Reviewer: ```suggestion

```

It feels to me like we can remove this sub-section, as it is already explained above.

> Primitive building block to handle event logic and event creation.
> It is imported via `import { rxActions } from '@rx-angular/state/actions';`

Reviewer: ```suggestion

```

It feels to me like we can remove this sub-section, as it is already explained above.

The action automatically creates an Observable and a callback function for imperative programming based on the types. It also has a new shorthand to tie the event changes to a side effect.

```typescript
import { rxActions } from '@rx-angular/state/actions';

@Component({...})
export class AnyComponent {
  ui = rxActions<{ name: string }>();
  name$ = this.ui.name$; // Observable<string> - listens to name changes
  emitName = this.ui.name; // (name: string) => void - emits name change
  sub = this.ui.onName(console.log) // () => void - when called stops the side effect execution
}
```

> ## Usage in component to handle UI interaction

Reviewer: ```suggestion
### Usage in component to handle UI interaction
```

This example shows how to create an action instance for the components template.
We can see how to apply behaviour onto the search input (debouncing the user input for 300 ms).

```typescript
import { rxActions } from '@rx-angular/state/actions';

type ListUi = { searchInput: string };

@Component({
  template: ` <input name="search" (change)="ui.searchInput($event)" /> `,
})
class ListComponent {
  protected ui = rxActions<ListUi>();
  // UI input with applied behaviour
  query$ = this.ui.searchInput$.pipe(debounceTime(300));
}
```

> ## Usage in a service to handle data fetching

Reviewer: ```suggestion
### Usage in a service to handle data fetching
```

In this example we see how to use an action instance to handle data fetching.
We can see how to apply behaviour onto the refresh calls (`exhaustMap` the HTTP requests).

```typescript
import { rxActions } from '@rx-angular/state/actions';

type Actions = { refresh: void };

@Injectable({ providedIn: 'root' })
export class MovieService {
  private actions = rxActions<Actions>();
  // data refresh with applied behaviour
  movies$ = this.actions.refresh$.pipe(exhaustMap((_) => getMovies()));

  refresh() {
    this.actions.refresh();
  }
}
```

> ## Handling side effects on event emission

Reviewer: ```suggestion
### Handling side effects on event emission
```

type EditUi = { save: void };

@Component({
  template: ` <button (click)="editUi.save($event)" /> `,
})
class ListComponent {
  protected editUi = rxActions<EditUi>();
  private saveObsEf = this.editUi.onSearchInput(
    (searchInput$) => searchInput$.pipe(debounceTime(300)), // apply behaviour
    (q: string) => this.state.search(q) // fire side effect
  );

  // It is also possible to use tap, however this is not recommended to do
  private saveObsEf2 = this.editUi.onSearchInput((query$) =>
    query$.pipe(
      debounceTime(300), // apply behaviour
      tap((q: string) => this.state.search(q)) // fire side effect
    )
  );
}
```

### Unsubscribe from events programmatically

```typescript
import { rxActions } from '@rx-angular/state/actions';

type EditUi = { save: void };

@Component({
  template: ` <button (click)="editUi.save($event)" /> `,
})
class ListComponent {
  protected editUi = rxActions<EditUi>();
  private saveObsEf = this.editUi.onSearchInput(
    (searchInput$) => searchInput$.pipe(debounceTime(300)), // apply behaviour
    (q: string) => this.state.search(q) // fire side effect
  );

  stopListenToClick() {
    this.saveObsEf();
  }
}
```

> ## Transform events in a central place

Reviewer: ```suggestion
### Transformations
```

}
```

## RxAngular Actions

This package helps to reduce code used to create composable action streams.
It mostly is used in combination with state management libs to handle user interaction and backend communication.

### Setup

The coalescing features can be used directly from the `cdk` package or indirectly through the `state` package.
To do so, install the `cdk` package and, if needed, the packages depending on it:

1. Install `@rx-angular/state`

```bash
npm i @rx-angular/state
# or
yarn add @rx-angular/state
```

### Basic usage in a Component

One of the ways `actions` can be used is to describe user input in a central place.
To do so we define the UI interactions a user can do e.g. if we have a search box in our UI, the events it produced could get named `searchInput`.click

The types for it could look like that:

````typescript
interface UI {
  searchInput: string;
}

In the class we can use actions like follows:

```typescript
class Component {

>   ui = actions<UI>();

Reviewer 2: What about using the new rxActions function in the docs? It simplifies the usage.

Author: Docs updates and ready for review

  searchQuery$ = this.ui.searchInput$.pipe(
    debounceTime(1500)
  );
  list$ = this.globalState.list$;

  constructor(
    private rxEffects: RxEffects,
    globalState: StateService
  ) {
    this.rxEffects.register(this.searchQuery$, this.globalState.refreshGenres);
  }
}
````

### Advanced usage

> 

Author: now i have the docs ready for review

> To ease this pain we can manage this login with `transforms`.

Reviewer: ```suggestion
By using the `transformations` API, we can preconfigure how our input events are mapped to actual values in a single place.
```

You can write your own transforms, leverage Browser APIs like `String` and `Boolean` or use the predefined functions.

> ### Use existing transform functions

Reviewer: ```suggestion
#### Use existing transform functions
```

The existing transform functions are:

- `preventDefault` - calls `preventDefault` on a passed event
- `stopPropagation`
- `preventDefaultStopPropagation`
- `eventValue`

Here we see how to use an action transforms. This concept is similar to Input transforms.
The logic is placed in a single place and transforms the event before emission.

```typescript
import { rxActions, eventValue } from '@rx-angular/state/actions';

type ListUi = { searchInput: string };

@Component({
  // takes a DOM Event
  template: `<input name="search" (change)="ui.searchInput($event)" />`,
})
class ListComponent {
  protected ui = rxActions<ListUi>(({ transform }) =>

>     transform({ searchInput: eventValue })

Reviewer: please add a comment on what `eventValue` actually does. For me it's not very clear

  );

  searchChange(query: string) {
    // takes a string
    this.ui.searchInput(query);
  }
}
```

> ## Use custom transform functions

Reviewer: ```suggestion
#### Use custom transform functions
```
